core/admin/mailu/utils.py

```diff
@@ -48,3 +63,79 @@ def init_app(self, app):
 
 # Data migrate
 migrate = flask_migrate.Migrate()
+
+
+# session store
+class NullSigner(object):
+    """NullSigner does not sign nor unsign"""
+    def __init__(self, *args, **kwargs):
+        pass
+    def sign(self, value):
+        """Signs the given string."""
+        return want_bytes(value)
+    def unsign(self, signed_value):
+        """Unsigns the given string."""
+        return want_bytes(signed_value)
+
+class KVSessionIntf(flask_kvsession.KVSessionInterface):
+    """ KVSession interface allowing to run int function on first access """
+    def __init__(self, app, init_fn=None):
+        if init_fn:
+            app.kvsession_init = init_fn
+        else:
+            self._first_run(None)
+    def _first_run(self, app):
+        if app:
+            app.kvsession_init()
+        self.open_session = super().open_session
+        self.save_session = super().save_session
+    def open_session(self, app, request):
+        self._first_run(app)
+        return super().open_session(app, request)
+    def save_session(self, app, session, response):
+        self._first_run(app)
+        return super().save_session(app, session, response)
+
+class KVSessionExt(flask_kvsession.KVSessionExtension):
+    """ Activates Flask-KVSession for an application. """
+    def init_kvstore(self, config):
+        """ Initialize kvstore - fallback to DictStore without REDIS_ADDRESS """
+        if addr := config.get('REDIS_ADDRESS'):
+            self.default_kvstore = RedisStore(redis.StrictRedis().from_url(f'redis://{addr}/3'))
+        else:
+            self.default_kvstore = DictStore()
```

Comment:
It may be worth logging something when this happens... Wouldn't it be better to check when the CLI is used and error out when it's not and there's no redis?

Reply:
Yes, this is not really a good behavior - but it is intended for debugging purposes only and not for the cli calls.
The cli actions are fine due to the "lazy" cleanup_sessions call: no session -> no cleanup -> no redis usage.

The DictStore is only used when REDIS_ADDRESS is explicitly set to the empty string.
When REDIS_ADDRESS is not set in the environment it will be filled by a hostname lookup.

Perhaps it would better to use an explicit "DEBUG" env variable. What do you think?

Reply:
I think it should error out if the address resolution fails in production... and warn that it's falling back to something else in DEBUG.

You could use this flag I guess:
https://flask.palletsprojects.com/en/master/api/#flask.Flask.debug

TBH, I'd also be fine with making redis connectivity mandatory if CLI doesn't need it.